src/dune_engine/scheduler.ml

```diff
@@ -759,45 +782,62 @@ let maybe_clear_screen ~config =
          ])
 
 let poll ?config ~once ~finally () =
-  let t = prepare ?config () in
-  let watcher = File_watcher.create () in
-  let block_waiting_for_changes () =
-    match Event.next () with
-    | Job_completed _ -> assert false
-    | Files_changed -> Continue
-    | Signal signal ->
-      got_signal signal;
-      Exit
-  in
-  let wait msg =
-    Console.Status_line.set_temporarily
-      (fun () ->
-        Some (Pp.seq msg (Pp.verbatim ", waiting for filesystem changes...")))
-      (fun () ->
-        let res = block_waiting_for_changes () in
-        maybe_clear_screen ~config;
-        res)
+  let t = prepare ?config ~polling:true () in
+  let finally () =
+    (* Re-create the throttle so that we don't keep old ivars from one run to
+       the next *)
+    t.job_throttle <- Fiber.Throttle.create (Fiber.Throttle.size t.job_throttle);
+    Fiber.return (finally ())
   in
+  let watcher = File_watcher.create () in
   let rec loop () =
-    let res = Run_once.run_and_cleanup t once in
-    finally ();
-    match res with
+    t.status <- Building;
+    let open Fiber.O in
+    let* res =
+      let on_error exn () =
+        match t.status with
+        | Building -> Report_error.report exn
+        | Cancelling_build -> ()
+        | Waiting_for_file_changes _ ->
+          (* We are inside a build, so we aren't waiting for a file change event *)
+          assert false
+      in
+      Fiber.fold_errors ~init:() ~on_error (fun () ->
+          Fiber.finalize ~finally once)
+    in
+    match t.status with
+    | Waiting_for_file_changes _ ->
+      (* We just finished a build, so there's no way this was set *)
+      assert false
+    | Cancelling_build -> loop ()
+    | Building ->
+      let message =
+        match res with
+        | Ok () -> Pp.tag User_message.Style.Success (Pp.verbatim "Success")
+        | Error _ -> Pp.tag User_message.Style.Error (Pp.verbatim "Had errors")
+      in
+      Console.Status_line.set
+        (Fun.const
+           (Some
+              (Pp.seq message
+                 (Pp.verbatim ", waiting for filesystem changes..."))));
+      let ivar = Fiber.Ivar.create () in
+      t.status <- Waiting_for_file_changes ivar;
+      let* () = Fiber.Ivar.read ivar in
+      maybe_clear_screen ~config;
+      loop ()
+  in
+  let exn, bt =
+    match Run_once.run_and_cleanup t loop with
     | Ok () ->
-      wait (Pp.tag User_message.Style.Success (Pp.verbatim "Success"))
-      |> after_wait
-    | Error Got_signal -> (Dune_util.Report_error.Already_reported, None)
-    | Error Never ->
-      wait (Pp.tag User_message.Style.Error (Pp.verbatim "Had errors"))
-      |> after_wait
-    | Error Files_changed -> loop ()
+      (* Polling mode is an infinite loop. We aren't going to terminate *)
```

Comment:
Shouldn't the type returned by `loop` be a `Nothing.t`, then, so we don't need this comment?

Reply:
Done.